### PR TITLE
[model, ci, docs] feat: bump transformers to 5.9.0

### PR DESCRIPTION
**Pull request by TimYangst**
### What does this PR do?

> Upgrades the pinned `transformers` version from `5.2.0` to `5.9.0` and
> regenerates all 16 patchgen configs (32 generated modeling files) against
> it.
>
> A version bump in a patchgen-based repo is not a one-line pin change: HF
> evolved several modeling files between 5.2 and 5.9, and the patches /
> patch configs had to be reconciled. This PR fixes every regression the
> upgrade surfaced (12 distinct root causes) so the full `tests/models/`
> suite passes clean on 5.9.0.
>
> Originally a 5.8.1 bump; review feedback on PR #774 was "just go to
> 5.9.0 in one step", and that's what this PR now does (the 5.8.1
> intermediate commits remain in the git history for traceability).

### Checklist Before Starting

- Search for relative PRs/issues and link here: N/A
- PR title follows `[{modules}] {type}: {description}` format

### Test

> All verified on an 8×A100-80GB worker with `transformers==5.9.0`:
>
> - `tests/models/` — **142 passed, 1 xfailed, 1 xpassed, 0 failed**.
> - `python -m veomni.patchgen.check_patchgen` — all 32 generated files
>   regenerate cleanly.
> - `make quality` — clean.

### API and Usage Example

> N/A — no VeOmni API change. `uv sync` now installs `transformers==5.9.0`
> via the `transformers-stable` default dependency group.

### Design & Code Changes

> Ten root causes were found and fixed; commits are split per cause:
>
> **From the 5.2 → 5.8.1 step (kept in history):**
>
> - **VLM `mm_token_type_ids`** — transformers v5 VLMs require
>   `mm_token_type_ids` (text=0 / image=1 / video=2) to compute multimodal
>   RoPE; `compute_3d_position_ids` / `get_rope_index` now raise without it.
>   Derive it from `input_ids` in the qwen2_vl / qwen3_vl / qwen3_vl_moe /
>   qwen3_5 / qwen3_5_moe forward patches and the `get_position_id`
>   data-pipeline helper; emit it from the VLM test fixture.
> - **Qwen2 `layer_types`** — HF 5.8 dropped `Qwen2DecoderLayer.attention_type`;
>   index the causal-mask map by `config.layer_types[i]` instead.
> - **Qwen3-MoE router dtype** — HF 5.8's `Qwen3MoeTopKRouter` casts the
>   top-k weights back to the raw-logits dtype; the patch kept fp32 on a
>   stale "no-op cast" assumption. Cast to `router_logits.dtype`.
> - **qwen3_5 cache classes** — HF 5.8 removed `Qwen3_5DynamicCache` /
>   `Qwen3_5MoeDynamicCache` (hybrid caches unified into `DynamicCache`);
>   the qwen3_5 / qwen3_5_moe configs imported them and silently failed to
>   regenerate. Switched to the base `Cache` type. Also switched the
>   `Qwen3_5MoeSparseMoeBlock` router-replay path to the qwen3_moe form,
>   as HF 5.8 fixed Qwen3.5-MoE's `TopKRouter` double-softmax quirk.
>
> **Additional changes for the 5.8.1 → 5.9.0 step:**
>
> - **`*VisionRotaryEmbedding.forward` signature** — 5.9 changed the
>   upstream signature from `(seqlen: int) -> (max_hw, dim//2)` to
>   `(position_ids: Tensor) -> (pos_ids.unsqueeze(-1) * inv_freq).flatten(1)`.
>   The patched `qwen3_vl` / `qwen3_5` `*VisionModel.rot_pos_emb` was
>   calling it with a Python int; rewrite to call
>   `self.rotary_pos_emb(pos_ids)` directly (the new API returns the
>   same `(N, dim)` shape as the old `freq_table[pos_ids].flatten(1)`
>   lookup, so the math is identical). NPU and MoE siblings inherit the
>   fix via name_map.
> - **`*VisionTransformerPretrainedModel.rot_pos_emb` deprecation** —
>   5.9 deprecated this shim (FutureWarning per forward, removed in
>   5.11) and internally calls `get_vision_position_ids(grid_thw, ...)`
>   which adds a `grid_thw.tolist()` sync. Replace the
>   `self.rot_pos_emb(grid_thw)` call in the patched
>   `qwen2_vl` / `qwen2_5_vl` / `qwen2_5_omni` / `qwen3_omni_moe` ViT
>   forwards with a host-driven `position_ids` build (mirroring upstream's
>   `get_vision_position_ids`) followed by `self.rotary_pos_emb(pos_ids)`.
>   This kills the deprecation warning, removes the internal `tolist()`
>   sync, and drops the now-dead `*.rot_pos_emb` /
>   `*VisionRotaryEmbedding.forward` allowlist entries.
> - **`F` re-import** — 5.9 dropped `import torch.nn.functional as F`
>   from the upstream `qwen2_vl` / `qwen2_5_vl` modeling files; the
>   patched ViT forward still calls `F.pad`. Add
>   `config.add_import("torch.nn.functional", alias="F", ...)` in both
>   patch configs.
> - **`create_causal_mask` / `create_sliding_window_causal_mask`
>   `cache_position`** — 5.9 dropped `cache_position` from these
>   constructors' signatures ("Deprecated and unused",
>   masking_utils.py:917). Strip it from the kwargs in the
>   `qwen2` / `qwen3_moe` / `qwen3_omni_moe` patches that build the mask
>   map.
> - **`fast_pos_embed_interpolate` dtype cast** — upstream 5.9's
>   `Qwen3OmniMoeVisionEncoder.forward` casts the interpolated positional
>   embedding back to `hidden_states.dtype` (the
>   `self.pos_embed(...) * bilinear_weights` product promotes bf16 ×
>   fp32 → fp32, which then breaks the bf16 LayerNorm in each vision
>   block with `expected scalar type Float but found BFloat16`). Mirror
>   that cast in the patched ViT forward.
> - **No-implicit-sync allowlists** — 5.9 moved several upstream sync
>   sites: introduced `*Model.get_vision_position_ids` (called from
>   `get_rope_index`, algorithm-essential), made
>   `*VisionRotaryEmbedding.forward` pure broadcast (no sync), and the
>   host-driven port above made `*.rot_pos_emb` paths unreachable in the
>   toy test. Refreshed `_ALLOWED_SYNCS` accordingly. Toned down the
>   `_PROD_PENDING_VIT_ROT_POS_EMB` wording to match.
> - **Omni audio encoder kwargs leak (`MODELING_BACKEND=veomni`)** — the
>   patched `Qwen2_5OmniAudioEncoder.forward` /
>   `Qwen3OmniMoeAudioEncoder.forward` were forwarding LM-level
>   `FlashAttentionKwargs` (`cu_seq_lens_q/k`, `max_length_q/k`) all the
>   way to `Qwen*OmniAudioAttention.forward`, which in 5.9 also passes
>   them explicitly → `got multiple values for keyword argument
>   'cu_seq_lens_q'`. Add the same `kwargs.pop` strip at the audio
>   encoder entry that we already do in the patched ViT forwards. Same
>   commit drops a vestigial `attention_mask=attention_mask` in the
>   qwen2_5_omni patched audio encoder→layer call (upstream 5.9's audio
>   encoder layer no longer accepts the kwarg).
> - **HF-backend test compatibility shim** — `tests/models/test_models_patch.py`
>   also exercises `MODELING_BACKEND=hf` (raw upstream) for HF↔VeOmni
>   numeric parity. Upstream qwen-family `Model.forward(**kwargs)` leaks
>   the same LM-level flash kwargs into `self.visual(...)` /
>   `self.audio_tower(...)`. In production we never hit this (VeOmni's
>   patched `Model.forward` builds clean per-modality kwarg dicts), but
>   the test does. Per `docs/transformers_v5/patchgen.md` we don't
>   monkey-patch upstream model code at runtime, so add the strip as a
>   test-local shim in `tests/models/test_models_patch.py` only — wraps
>   the upstream `*VisionEncoder.forward` / `*AudioEncoder.forward` to
>   pop the leaked keys at the encoder entry. No-op on the patched
>   VeOmni classes (different Python objects). Drop when upstream stops
>   leaking.
> - **Patchgen `Based on:` headers** — every regen file's
>   `# Based on: transformers==X.Y.Z` banner bumped to 5.9.0.
>
> Plus: `pyproject.toml` pin + `uv.lock`, and `5.2.0 → 5.9.0`
> version-string updates across `constraints.md`, `uv.md`, the skill
> docs, install / FAQ docs and `patchgen.md`; the
> `is_transformers_version_greater_or_equal_to` guardrail wording in
> `constraints.md` tightened to forbid version gates for `<= 5.9.0`.

### Known follow-ups (intentionally out of scope)

> Both are 5.11-deadline items — the calls still work in 5.9, just emit
> FutureWarnings:
>
> - `*VisionEncoder.get_window_index` — still called from the patched
>   `qwen2_5_vl` / `qwen2_5_omni` ViT forwards (the precompute consumer
>   path already bypasses it, but the fallback path does not).
> - `*VisionEncoder.fast_pos_embed_interpolate` — still called from the
>   patched `qwen3_omni_moe` ViT forward.

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- [x] Applied pre-commit checks
- [x] Added/updated documentation
- [x] Added tests to CI workflow (or explained why not feasible) — existing
      `tests/models/` covers the change; the VLM test fixture was updated
      to emit `mm_token_type_ids` in the 5.8.1 step.

### CODE REVIEW

**Comment by Coach257** [comment]
Maybe bump to 5.9.0 https://github.com/huggingface/transformers/releases/tag/v5.9.0

**Comment by TimYangst** [comment]
> Maybe bump to 5.9.0 https://github.com/huggingface/transformers/releases/tag/v5.9.0

Sounds good. Since this is a relatively large change, I’ll spend some time re-adapting and validating the integration.

**Comment by TimYangst** [comment]
@gemini-code-assist  help to take a review again, please

File: veomni/models/transformers/qwen3_5/qwen3_5_gpu_patch_gen_config.py [hunk 1]
```diff
@@ -306,7 +305,7 @@ def qwen3_5_gated_deltanet_get_local_conv1d_weight(
 def qwen3_5_gated_deltanet_forward_patched(
     self,
     hidden_states: torch.Tensor,
-    cache_params: Qwen3_5DynamicCache | None = None,
+    cache_params: Cache | None = None,
     cache_position: torch.LongTensor | None = None,
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/models/transformers/qwen3_5/qwen3_5_gpu_patch_gen_config.py:316]: Outdated

File: veomni/models/transformers/glm_moe_dsa/generated/patched_modeling_glm_moe_dsa_gpu.py [hunk 2]
```diff
@@ -222,7 +222,7 @@ def forward(
 
         # q·k^T per head: [B, S, H, D] @ [B, T, D]^T → [B, S, H, T]
         scores = torch.einsum("bshd,btd->bsht", q.float(), k_cached.float()) * self.softmax_scale
-
+        scores = F.relu(scores)
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/models/transformers/glm_moe_dsa/generated/patched_modeling_glm_moe_dsa_gpu.py:225]: Outdated

File: veomni/models/transformers/qwen3_vl/generated/patched_modeling_qwen3_vl_gpu.diff [hunk 3]
```diff
@@ -386,8 +411,10 @@
      def forward(
          self,
          hidden_states: torch.Tensor,
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/models/transformers/qwen3_vl/generated/patched_modeling_qwen3_vl_gpu.diff:599]: Fixed

File: veomni/models/transformers/qwen3_vl/generated/patched_modeling_qwen3_vl_npu.diff [hunk 4]
```diff
@@ -368,7 +386,7 @@
              attn_outputs = [
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/models/transformers/qwen3_vl/generated/patched_modeling_qwen3_vl_npu.diff:454]: Outdated

File: veomni/models/transformers/qwen2_vl/generated/patched_modeling_qwen2_vl_gpu.py [hunk 5]
```diff
@@ -1430,12 +1413,13 @@ def compute_3d_position_ids(
                 position_ids = torch.arange(past_key_values_length, past_key_values_length + seq_length)
                 position_ids = position_ids.view(1, 1, -1).expand(3, batch_size, -1).to(inputs_embeds.device)
             delta = self.rope_deltas.repeat_interleave(batch_size // self.rope_deltas.shape[0], dim=0)
-            position_ids = position_ids + delta.to(device=position_ids.device)
+            position_ids = position_ids + delta.to(device=inputs_embeds.device)
         else:
-            # Can't build correct 3D positions. Let the model infer it from `cache_position`
+            # Can't build correct 3D positions. Let the model infer it
             position_ids = None
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/models/transformers/qwen2_vl/generated/patched_modeling_qwen2_vl_gpu.py:1439]: Upstream HF 5.9 Qwen2VLTextModel.forward already handles position_ids=None with exactly the fallback Gemini is suggesting.

**Review by Luosuu — APPROVED** [review]

**Comment by Luosuu** [comment]
@TimYangst please resolve the conflict

**Review by Coach257 — APPROVED** [review]
